### ISSUE TITLE
Nodetool repair stuck for several hours after another node was stopped and terminated while in parallel schema changes were running (Create SI)

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylladb#20106]
## Packages

Scylla version: `6.1.0-20240804.abbf0b24a60c` with build-id `374d38ae5c8f543c472d3920542bbb543be25d09`

Kernel Version: `6.8.0-1012-aws`

## Issue description

- [x] This issue looks like a regression.
- [ ] It is unknown if this issue is a regression.

During test in parallel running two operations: topology and schema changes. Cluster has multidc (2 dc) configuration.

Topology operation: Stop node, terminate it and then run repair on each node in cluster and remove the node and then replace removed node with new one
Schema operations: Adddrop Column, Create Index.

While schema operations has been running, node 13 was stopped and then terminated. After that `nodetool repair` was running on each node.
When repair started on node8: 
```
2024-08-07T19:22:05.739+00:00 parallel-topology-schema-changes-mu-db-node-0b612412-8     !INFO | scylla[6428]:  [shard 0:strm] repair - repair[46cbc0fa-cb2a-4d29-aeb4-6e28cba9ed91]: starting user-requested repai
r for keyspace system_auth, repair id 6, options {"trace": "false", "primaryRange": "false", "jobThreads": "1", "incremental": "false", "parallelism": "parallel"}
2024-08-07T19:22:05.739+00:00 parallel-topology-schema-changes-mu-db-node-0b612412-8     !INFO | scylla[6428]:  [shard 0:strm] repair - repair[46cbc0fa-cb2a-4d29-aeb4-6e28cba9ed91]: completed successfully: no ta
bles to repair
2024-08-07T19:22:05.739+00:00 parallel-topology-schema-changes-mu-db-node-0b612412-8     !INFO | scylla[6428]:  [shard 0:strm] repair - repair[b50a906d-bd33-4aae-951c-25f227ebba3f]: starting user-requested repai
r for keyspace keyspace1, repair id 7, options {"trace": "false", "primaryRange": "false", "jobThreads": "1", "incremental": "false", "parallelism": "parallel"}
2024-08-07T19:22:05.739+00:00 parallel-topology-schema-changes-mu-db-node-0b612412-8     !INFO | scylla[6428]:  [shard 0:strm] repair - repair[b50a906d-bd33-4aae-951c-25f227ebba3f] Table keyspace1.standard2 has 
tablet transitions, waiting for topology to quiesce
```
Looks like command and process was stucked on the node 8.

## Impact

Command and process is stucked, could affect on data and cluster avialability

## How frequently does it reproduce?

rarely

## Installation details

Cluster size: 12 nodes (i3en.2xlarge)

Scylla Nodes used in this run:
- parallel-topology-schema-changes-mu-db-node-0b612412-1 (108.129.93.80 | 2a05:d018:12e3:f000:8905:b723:136f:5b20) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-10 (3.8.138.24 | 2a05:d01c:0964:7d01:9e37:518f:c729:6bac) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-11 (18.134.12.117 | 2a05:d01c:0964:7d02:4243:5276:3f4a:28e1) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-12 (18.168.198.32 | 2a05:d01c:0964:7d02:4ebe:795d:f11d:d12c) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-13 (34.246.176.30 | 2a05:d018:12e3:f002:36b8:0d8f:4033:7e79) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-2 (34.245.56.225 | 2a05:d018:12e3:f000:fb0f:a94d:1624:de16) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-3 (3.249.5.244 | 2a05:d018:12e3:f001:6f9c:f987:14e9:d21c) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-4 (52.212.94.78 | 2a05:d018:12e3:f001:5235:1a46:bc52:638e) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-5 (18.201.74.209 | 2a05:d018:12e3:f002:af0c:1f8c:6f99:f158) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-6 (34.255.215.192 | 2a05:d018:12e3:f002:f24f:e9b3:ba6f:5cba) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-7 (35.176.224.142 | 2a05:d01c:0964:7d00:61a3:e037:9e2e:37e8) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-8 (35.176.164.26 | 2a05:d01c:0964:7d00:4c73:c84e:1286:4738) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-0b612412-9 (35.178.151.47 | 2a05:d01c:0964:7d01:bc14:b2fc:7bea:a57c) (shards: 7)

OS / Image: `ami-04d60427a034a2322 ami-029d1e3ee17a49467` (aws: undefined_region)

Test: `longevity-multidc-schema-topology-changes-12h-test`
Test id: `0b612412-560e-465e-a58a-e675376f7894`
Test name: `scylla-6.1/tier1/longevity-multidc-schema-topology-changes-12h-test`
Test method: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):

- [longevity-multidc-parallel-topology-schema-changes-12h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/3c2733ec45998d5348209f61fa177972a1afbfba/test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 0b612412-560e-465e-a58a-e675376f7894`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=0b612412-560e-465e-a58a-e675376f7894)
- Show all stored logs command: `$ hydra investigate show-logs 0b612412-560e-465e-a58a-e675376f7894`

## Logs:
- **db-cluster-0b612412.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/db-cluster-0b612412.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/db-cluster-0b612412.tar.gz)
- **sct-runner-events-0b612412.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/sct-runner-events-0b612412.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/sct-runner-events-0b612412.tar.gz)
- **sct-0b612412.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/sct-0b612412.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/sct-0b612412.log.tar.gz)
- **loader-set-0b612412.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/loader-set-0b612412.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/loader-set-0b612412.tar.gz)
- **monitor-set-0b612412.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/monitor-set-0b612412.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0b612412-560e-465e-a58a-e675376f7894/20240808_064343/monitor-set-0b612412.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-6.1/job/tier1/job/longevity-multidc-schema-topology-changes-12h-test/4/)
[Argus](https://argus.scylladb.com/test/a5b48503-1d3d-463d-a870-ae0794a024cf/runs?additionalRuns[]=0b612412-560e-465e-a58a-e675376f7894)
</details>

**Comment by temichus** [comment]
@mykaul, the new issue for 6.1

**Comment by mykaul** [comment]
> Table keyspace1.standard2 has 
> tablet transitions, waiting for topology to quiesce

@tgrabiec - can you take a look (based on the above log line, might be related to serialization of repair and migrations?)

**Comment by tgrabiec** [comment]
The repair is blocked by tablet migration, and tablet migration is stuck due to topology barrier failing and migration not being able to make progress:

```
Aug 07 19:05:03.735477 parallel-topology-schema-changes-mu-db-node-0b612412-1 scylla[6390]:  [shard 0: gms] raft_topology - drain rpc failed, proceed to fence old writes: std::runtime_error (raft topology: exec_global_command(barrier_and_
drain) failed with seastar::rpc::closed_error (connection is closed))
Aug 07 19:05:04.020015 parallel-topology-schema-changes-mu-db-node-0b612412-1 scylla[6390]:  [shard 0: gms] raft_topology - barrier for tablet 171ed8c0-54eb-11ef-b4b0-6ff4c217f052:15 failed: seastar::broken_promise (broken promise)
Aug 07 19:05:04.020060 parallel-topology-schema-changes-mu-db-node-0b612412-1 scylla[6390]:  [shard 0: gms] raft_topology - topology change coordinator fiber got error std::runtime_error (raft topology: exec_global_command(barrier) failed with seastar::rpc::closed_error (connection is closed))
```

Sadly, we don't log which node failed to respond. There is an issue about it: https://github.com/scylladb/scylladb/issues/16923. 

But it's probably the node which was terminated: 683e72eb-9699-4f70-b9c3-c24c08b281ef

The way out of this situation is to mark the node as dead by calling removenode on it, which will allow migrations to progress and unblock repair.

**Comment by mykaul** [comment]
Thanks @tgrabiec - so we have a way out here, thus reducing severity. We need to see how we do not fail on this in SCT tests :-/

**Comment by tgrabiec** [comment]
In this case, if repair wasn't stuck on migration, it would probably fail due to replica being down.

I don't think schema changes are relevant to this scenario, maybe except slowing some things down and affecting timing of migration.

**Comment by tgrabiec** [comment]
@aleksbykov So I think it's a problem on the test side. With tablets, you need to call removenode first in order to be able to run repair.

**Comment by fruch** [comment]
> @aleksbykov So I think it's a problem on the test side. With tablets, you need to call removenode first in order to be able to run repair.

is that documented anywhere ? 

@pehala please see if there are any change like that related to tablets is needed, someone would need to chase them.

**Comment by mykaul** [comment]
@pehala ?

**Comment by fruch** [comment]
@tgrabiec 

I've read this thread again

this nemesis was introduced in 2020: https://github.com/scylladb/scylla-cluster-tests/pull/2331

and it follows the instructions to the letter:
https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/remove-node#removing-an-unavailable-node

especially this one:

> - Run a full cluster repair **before** nodetool removenode, so all existing replicas have the most up-to-date data.

the the next instruction wasn't there in 2020:

>  - In the case of node failures during the removenode operation, re-run repair before running nodetool removenode (not required when [Repair Based Node Operations (RBNO)](https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/repair-based-node-operation.html) for removenode is enabled).

so if RBNO is used for removenode, we shouldn't do a full cluster repair ?

**Comment by denesb** [comment]
> so if RBNO is used for removenode, we shouldn't do a full cluster repair ?

If you use RBNO, you don't need extra repair, repair happens during the node op, whatever node op that may be.

**Comment by paszkow** [comment]
No update since 2024. Closing. We can re-open once we get on it again.
